File: algorithm/Java_Algorithm/src/f_search/Transpose.java
package f_search;

public class Transpose extends SinglyLinkedList {
	public Transpose() {
	}

	private Node transpose(int target) {
		Node current = super.head;
		Node match = null;
		Node pprevious = null;
		Node previous = null;

		while (current != null) {
			if (current.getValue() == target) {
				match = current;
				if (previous != null) {
					if (pprevious != null) {
						pprevious.setNextNode(current);
					} else {
						super.head = current;
					}
					previous.setNextNode(current.getNextNode());
					current.setNextNode(previous);
				}
				break;
			} else {
				if (previous != null) {
					pprevious = previous;
				}
				previous = current;
				current = current.getNextNode();
			}

		}

		return match;
	}
	
	

	public static void main(String args[]) {
		Transpose t = new Transpose();
		t.add(1);
		t.add(2);
		t.add(3);
		t.add(4);
		t.add(5);
		t.add(6);
		t.add(7);
		
		Node node = t.transpose(6);
		
		
		
		t.printTree();
		
		
		
	}
	
}

File: algorithm/[book]c_algorithm/src/f.6-3_transpose/Transpose.c
/*
 * Transpose.c
 *
 *  Created on: 2019. 1. 5.
 *      Author: otrodevym
 */
#include "Transpose.h"

Node* SLL_Transpose(Node** Head, int Target) {
	Node* Current = (*Head);
	Node* Match = NULL;
	Node* PPrevious = NULL;
	Node* Previous = NULL;

	while (Current != NULL) {
		if (Current->Data == Target) {
			Match = Current;
			if (Previous != NULL) {
				if (PPrevious != NULL) {
					PPrevious->NextNode = Current;

				} else {
					(*Head) = Current;
				}
				Previous->NextNode = Current->NextNode;

				Current->NextNode = Previous;
			}
			break;

		} else {
			if(Previous != NULL){
				PPrevious = Previous;
			}
				Previous = Current;
				Current = Current->NextNode;
		}
	}
	return Match;
}
